src/folding/nova/decider.rs

```diff
@@ -0,0 +1,647 @@
+        })?;
+        let z_i = Vec::<FpVar<CF1<C1>>>::new_witness(cs.clone(), || {
+            Ok(self.z_i.unwrap_or(vec![CF1::<C1>::zero()]))
+        })?;
+
+        let u_dummy_native = CommittedInstance::<C1>::dummy(1);
+        let w_dummy_native = Witness::<C1>::new(
+            vec![C1::ScalarField::zero(); self.r1cs.A.n_cols - 2 /* (2=1+1, since u_i.x.len=1) */],
+            self.E_len,
+        );
+
+        let u_i = CommittedInstanceVar::<C1>::new_witness(cs.clone(), || {
+            Ok(self.u_i.unwrap_or(u_dummy_native.clone()))
+        })?;
+        let w_i = WitnessVar::<C1>::new_witness(cs.clone(), || {
+            Ok(self.w_i.unwrap_or(w_dummy_native.clone()))
+        })?;
+        let U_i = CommittedInstanceVar::<C1>::new_witness(cs.clone(), || {
+            Ok(self.U_i.unwrap_or(u_dummy_native.clone()))
+        })?;
+        let W_i = WitnessVar::<C1>::new_witness(cs.clone(), || {
+            Ok(self.W_i.unwrap_or(w_dummy_native.clone()))
+        })?;
+
+        let crh_params = CRHParametersVar::<C1::ScalarField>::new_constant(
+            cs.clone(),
+            self.poseidon_config.clone(),
+        )?;
+
+        // 1. check RelaxedR1CS of u_i
+        let z_u: Vec<FpVar<CF1<C1>>> = [
+            vec![FpVar::<CF1<C1>>::one()],
+            u_i.x.to_vec(),
+            w_i.W.to_vec(),
+        ]
+        .concat();
+        RelaxedR1CSGadget::<C1::ScalarField, CF1<C1>, FpVar<CF1<C1>>>::check(
+            r1cs.clone(),
+            w_i.E,
+            u_i.u.clone(),
+            z_u,
+        )?;
+
+        // 2. check RelaxedR1CS of U_i
+        let z_U: Vec<FpVar<CF1<C1>>> =
+            [vec![U_i.u.clone()], U_i.x.to_vec(), W_i.W.to_vec()].concat();
+        RelaxedR1CSGadget::<C1::ScalarField, CF1<C1>, FpVar<CF1<C1>>>::check(
+            r1cs,
+            W_i.E,
+            U_i.u.clone(),
+            z_U,
+        )?;
+
+        // 3. u_i.cmE==cm(0), u_i.u==1
+        let zero_x = NonNativeFieldVar::<C1::BaseField, C1::ScalarField>::new_constant(
+            cs.clone(),
+            C1::BaseField::zero(),
+        )?
+        .to_constraint_field()?;
+        let zero_y = NonNativeFieldVar::<C1::BaseField, C1::ScalarField>::new_constant(
+            cs.clone(),
+            C1::BaseField::one(),
+        )?
+        .to_constraint_field()?;
+        (u_i.cmE.x.is_eq(&zero_x)?).enforce_equal(&Boolean::TRUE)?;
+        (u_i.cmE.y.is_eq(&zero_y)?).enforce_equal(&Boolean::TRUE)?;
+        (u_i.u.is_one()?).enforce_equal(&Boolean::TRUE)?;
+
+        // 4. u_i.x == H(i, z_0, z_i, U_i)
+        let u_i_x = U_i
+            .clone()
+            .hash(&crh_params, i.clone(), z_0.clone(), z_i.clone())?;
+        (u_i.x[0]).enforce_equal(&u_i_x)?;
+
+        // The following two checks (and their respective allocations) are disabled for normal
+        // tests since they take ~24.5M constraints and would take several minutes (and RAM) to run
+        // the test
+        #[cfg(not(test))]
```

Comment:
We maybe should do this differently and use a feature or simply remove it. 
Basically because we will benchmark within an integration test the real workflow. And this will prevent us to do so. 


I'd simply remove this (at the end, we aren't testing this function anyways in this PR).
And otherwise, let's do a feature rather than this.

Reply:
it is actually being executed in the test at the end of this file 
https://github.com/privacy-scaling-explorations/folding-schemes/tree/9e20fbc385daf30d9ef7e4f3f0f76fb771daea32/src/folding/nova/decider.rs#L643 , that's why I was 'disabling' it, because it takes too much ram and time to process.
Will explore the feature approach that you mention.

Reply:
Maybe what we can do is add a feature that disables the expensive part for the testing. And just leave the other part.

Which is what you made :sweat_smile: 

Yes, I agree this might be the best way to go, we b default run always the entire thing in production but we don't run the expensive part in the CI tests.

And then we can setup a machine that runs each week or so the circuit for the full decider in our worker.

Reply:
Update: in a call with @CPerezz we agreed to keep it as it is, and in the next days we will prepare a kind of weekly test to be run in a powerful machine in release mode executing the full circuit.